padertorch/contrib/examples/tasnet/train.py

```diff
@@ -0,0 +1,309 @@
+    next to it. If a Makefile already exists, it does not do anything.
+    """
+
+    MAKEFILE_TEMPLATE = """
+    SHELL := /bin/bash
+    MODEL_PATH := $(shell pwd)
+
+    export OMP_NUM_THREADS=1
+    export MKL_NUM_THREADS=1
+
+    train:
+    \tpython -m {main_python_path} with config.json
+
+    ccsalloc:
+    \tccsalloc \\
+    \t\t--notifyuser=awe \\
+    \t\t--res=rset=1:ncpus=4:gtx1080=1:ompthreads=1 \\
+    \t\t--time=100h \\
+    \t\t--join \\
+    \t\t--stdout=stdout \\
+    \t\t--tracefile=%x.%reqid.trace \\
+    \t\t-N train_{nickname} \\
+    \t\tpython -m {main_python_path} with config.json
+
+    evaluate:
+    \tpython -m {eval_python_path} init with model_path=$(MODEL_PATH)
+    """
+
+    experiment_dir = Path(_config['trainer']['storage_dir'])
+    makefile_path = Path(experiment_dir) / "Makefile"
+
+    if not makefile_path.exists():
+        config_path = experiment_dir / "config.json"
+
+        pb.io.dump_json(_config, config_path)
+
+        makefile_path.write_text(MAKEFILE_TEMPLATE.format(
+            main_python_path=pt.configurable.resolve_main_python_path(),
+            experiment_dir=experiment_dir,
+            nickname=nickname,
+            eval_python_path='.'.join(
+                pt.configurable.resolve_main_python_path().split('.')[:-1]
+            ) + '.evaluate',
+            model_path=Path(experiment_dir)
+        ))
+
+
+@ex.command
+def init(_config, _run):
+    """Create a storage dir, write Makefile. Do not start any training."""
+    sacred.commands.print_config(_run)
+    dump_config_and_makefile()
+
+    print()
+    print('Initialized storage dir. Now run these commands:')
+    print(f"cd {_config['trainer']['storage_dir']}")
+    print(f"make train")
+    print()
+    print('or')
+    print()
+    print(f"cd {_config['trainer']['storage_dir']}")
+    print('make ccsalloc')
+
+
+@ex.capture
+def prepare_and_train(_run, _log, trainer, train_dataset, validate_dataset,
+                      lr_scheduler_step, lr_scheduler_gamma,
+                      load_model_from, database_json):
+    trainer = pt.Trainer.from_config(trainer)
+    checkpoint_path = trainer.checkpoint_dir / 'ckpt_latest.pth'
+
+    if load_model_from is not None and not checkpoint_path.is_file():
+        _log.info(f'Loading model weights from {load_model_from}')
+        checkpoint = torch.load(load_model_from)
+        trainer.model.load_state_dict(checkpoint['model'])
+
+    db = JsonDatabase(database_json)
+
+    # Perform a test run to check if everything works
+    trainer.test_run(
+        prepare_iterable_captured(db, train_dataset),
+        prepare_iterable_captured(db, validate_dataset),
+    )
+
+    # Register hooks and start the actual training
+    trainer.register_validation_hook(
+        prepare_iterable_captured(db, validate_dataset)
+    )
+
+    # Learning rate scheduler
+    trainer.register_hook(pt.train.hooks.LRSchedulerHook(
```

Comment:
I think @jensheit removed the `pt.train.hooks.LRSchedulerHook` in favor of the BackOffValidationHook

Reply:
Yes, and the BackOffValidationHook should be added with register_validation_hook()

Reply:
But those two do different things. This one is as close to the original paper as possible. I don't know how they compare in terms of training behavior, but the backoff validation hook might be preferable

Reply:
The `pt.train.hooks.LRSchedulerHook` is still available.
In #11 only LRSchedulers in `pt.train.optimizer` were deleted, `pt.train.hooks.LRSchedulerHook` is an alternative to the LRSchedulers in `pt.train.optimizer` and is now the recommended one, when no backoff is used/needed.